### PR TITLE
feat(core): add api logging provider

### PR DESCRIPTION
**Pull request by iartemiev**
#### Description of changes
**This will be a tagged release (`@api-logging`) and will be undocumented at first. Docs will follow once it is feature-complete.**

Enables sending library and application logs to an API endpoint.

### Example use

Send all library logs to the provided endpoint:
```js
import { Logger } from 'aws-amplify';

Logger.addPluggable(
  new APILoggingProvider({
    endpoint: 'http://localhost:3000',
  })
);
```

#### Description of how you validated changes
* React and RN sample apps
* Unit tests 

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions) 
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/Logger/ConsoleLogger.ts [hunk 1]
```diff
@@ -77,6 +82,43 @@ export class ConsoleLogger implements Logger {
...
+			const event = {
+				data,
+				level: type,
+				message: strMessage,
+				source: this.name,
+				timestamp: Date.now(),
+			};
+
+			return event;
+		};
```
Comment (elorzafe) [inline packages/core/src/Logger/ConsoleLogger.ts:114]: Maybe take that function out to make the code easy to read?

File: packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+export const truncateOversizedEvent = (event: InputLogEvent): InputLogEvent => {
+	const { timestamp, message } = event;
+
+	try {
+		const messageJson = JSON.parse(message);
+
+		const truncatedObj = {
+			level: messageJson.level,
+			class: messageJson.class,
+			message: messageJson.message.substring(0, 500),
```
Comment (elorzafe) [inline packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts:32]: Where is this `500` coming from?

File: packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		};
+
+		if (messageJson.data != null) {
+			truncatedObj[
+				'data'
+			] = `OBJECT SIZE EXCEEDS CLOUDWATCH EVENT LIMIT. Truncated: ${JSON.stringify(
+				messageJson.data
+			).substring(0, 500)}`;
```
Comment (elorzafe) [inline packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts:40]: Why `500`?

File: packages/core/src/Providers/APIProvider/APILoggingProvider.ts [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+	): GenericLogEvent[] | InputLogEvent[] {
+		if (this.config.eventFormat === EVENT_FORMAT.CLOUDWATCH) {
+			return cloudWatchLogEventBatch(buffer as InputLogEvent[]);
+		}
+
+		return this.genericLogEventBatch(buffer as GenericLogEvent[]);
+	}
+
+	private genericLogEventBatch(buffer: GenericLogEvent[]): GenericLogEvent[] {
+		// TODO: Add logic to limit batch + headers to 6MB; the Lambda payload limit
```
Comment (elorzafe) [inline packages/core/src/Providers/APIProvider/APILoggingProvider.ts:259]: Is this going to happen on this iteration?

File: packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+		});
+
+		return {
+			timestamp,
+			message: truncated,
+		};
+	}
+};
+
+export const cloudWatchLogEventBatch = (
```
Comment (elorzafe) [inline packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts:64]: I am not a fan of changing the input directly on this method, I would remove those once are sent. What do you think?

**Review by elorzafe — COMMENTED** [review]
A few comments we can chat more.

Overall looks very good

Thanks @iartemiev !

File: packages/core/src/Providers/APIProvider/loggerConnectivity.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	socketDisconnected() {
+		if (this.observer && typeof this.observer.next === 'function') {
+			this.observer.next({ online: false });
+
+			this.timeout = setTimeout(() => {
+				const observerResult = { ...this.connectionStatus };
+				this.observer.next(observerResult);
+			}, RECONNECTING_IN);
+		}
+	}
```
[the remark this answers is not in the document]
Comment (iartemiev) [reply packages/core/src/Providers/APIProvider/loggerConnectivity.ts]: Removed. I don't think it's used

File: packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts (hunk 2 above)
Comment (iartemiev) [reply packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts]: Bumped to 5000 and moved to a constant here: https://github.com/aws-amplify/amplify-js/pull/9689/files#diff-44e36fcf66f06835ef49564165f7be290cf16b238efbdaece2bfa3287b0557a1R10

Went with 5000 to cover worst case where both `message` and `data` are over the limit. 10k combined characters will keep it under the CW max event size limit

File: packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts (hunk 3 above)
Comment (iartemiev) [reply packages/core/src/Providers/APIProvider/CloudWatchEventFormatter.ts]: Responded above

File: packages/core/src/Providers/APIProvider/APILoggingProvider.ts (hunk 4 above)
Comment (iartemiev) [reply packages/core/src/Providers/APIProvider/APILoggingProvider.ts]: Yep, added

**Review by elorzafe — APPROVED** [review]
LGTM, thanks @iartemiev 🌮 🎉

**Pull request merged**